xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESink.cpp

```diff
@@ -728,6 +728,114 @@ void CActiveAESink::PrintSinks(std::string& driver)
   }
 }
 
+std::string CActiveAESink::ValidateOuputDevice(const std::string& device, bool passthrough)
+{
+  if (m_sinkInfoList.empty())
+    return {};
+
+  const AESinkDevice dev = CAESinkFactory::ParseDevice(device);
+
+  // find exact match of deviceName in same driver
+  if (!dev.driver.empty() && !dev.name.empty())
+  {
+    for (const auto& sink : m_sinkInfoList)
+    {
+      if (sink.m_sinkName != dev.driver)
+        continue;
+
+      for (const auto& d : sink.m_deviceInfoList)
+      {
+        if (passthrough && (d.m_deviceType == AE_DEVTYPE_PCM || d.m_onlyPCM))
+          continue;
+
+        if (!passthrough && d.m_onlyPassthrough)
+          continue;
+
+        if (d.m_deviceName == dev.name)
+          return d.ToDeviceString(sink.m_sinkName);
+      }
+    }
+  }
+
+  // find same friendly name on other device in same driver
+  if (!dev.driver.empty() && !dev.friendlyName.empty())
+  {
+    for (const auto& sink : m_sinkInfoList)
+    {
+      if (sink.m_sinkName != dev.driver)
+        continue;
+
+      for (const auto& d : sink.m_deviceInfoList)
+      {
+        if (passthrough && (d.m_deviceType == AE_DEVTYPE_PCM || d.m_onlyPCM))
+          continue;
+
+        if (!passthrough && d.m_onlyPassthrough)
+          continue;
+
+        if (d.GetFriendlyName() == dev.friendlyName)
+          return d.ToDeviceString(sink.m_sinkName);
+      }
+    }
+  }
+
+  std::string firstDevice;
+
+  // find default device of same driver or first device of same driver
+  if (!dev.driver.empty())
+  {
+    for (const auto& sink : m_sinkInfoList)
+    {
+      if (sink.m_sinkName != dev.driver)
+        continue;
+
+      for (const auto& d : sink.m_deviceInfoList)
+      {
+        if (passthrough && (d.m_deviceType == AE_DEVTYPE_PCM || d.m_onlyPCM))
+          continue;
+
+        if (!passthrough && d.m_onlyPassthrough)
+          continue;
+
+        if (firstDevice.empty())
+          firstDevice = d.ToDeviceString(sink.m_sinkName);
+
+        if (d.m_deviceName.find("default") != std::string::npos)
+          return d.ToDeviceString(sink.m_sinkName);
+      }
+
+      if (!firstDevice.empty())
+        break;
+    }
+  }
+
+  // return first device of same driver
+  if (!firstDevice.empty())
+    return firstDevice;
+
+  // find the default of any driver
+  for (const auto& sink : m_sinkInfoList)
+  {
+    for (const auto& d : sink.m_deviceInfoList)
+    {
+      if (passthrough && (d.m_deviceType == AE_DEVTYPE_PCM || d.m_onlyPCM))
+        continue;
+
+      if (!passthrough && d.m_onlyPassthrough)
+        continue;
+
+      if (d.m_deviceName.find("default") != std::string::npos)
+        return d.ToDeviceString(sink.m_sinkName);
+    }
+  }
+
+  if (m_sinkInfoList[0].m_deviceInfoList.empty())
```

Comment:
At this point, should probably look for any device for any driver. driver 0 may not have devices but driver 1 may.
If nothing is found then return {}
Could be integrated with the search of a default device for any driver in the previous block, same way it's done for the matching driver a bit earlier?

Reply:
done